std/hash/keccak256/keccak.go

```diff
@@ -0,0 +1,120 @@
+package keccak
+
+import (
+	"github.com/consensys/gnark/frontend"
+	keccakf2 "github.com/consensys/gnark/std/permutation/keccakf"
+)
+
+const Size = 256 / 8
+
+const BlockSize = 1600/8 - Size*2
+
+// Keccak256 implements hash.Hash
+// variable == single byte
+type Keccak256 struct {
+	a      [25]keccakf2.Xuint64
+	buf    [200]keccakf2.Xuint8
+	dsbyte keccakf2.Xuint8
+	len    int
+	size   int
+	api    frontend.API
+	uapi64 *keccakf2.Uint64api
+	uapi8  *keccakf2.Uint8api
+}
+
+func (h *Keccak256) Api() frontend.API {
+	return h.api
+}
+
+func NewKeccak256(api frontend.API) Keccak256 {
+	return Keccak256{
+		dsbyte: keccakf2.ConstUint8(0x01),
+		size:   256 / 8,
+		api:    api,
+		uapi64: keccakf2.NewUint64API(api),
+		uapi8:  keccakf2.NewUint8API(api),
+	}
+}
+
+func (h *Keccak256) Size() int      { return h.size }
+func (h *Keccak256) BlockSize() int { return BlockSize }
+
+func (h *Keccak256) Reset() {
+	h.a = [25]keccakf2.Xuint64{}
+	for i := range h.a {
+		h.a[i] = keccakf2.ConstUint64(0)
+	}
+	h.buf = [200]keccakf2.Xuint8{}
+	h.len = 0
+}
+
+func (h *Keccak256) Write(data ...frontend.Variable) {
+	bs := h.BlockSize()
+
+	in := make([]keccakf2.Xuint8, len(data))
+	for i := range data {
+		in[i] = h.uapi8.AsUint8(data[i])
+	}
+
+	for len(in) > 0 {
```

Comment:
maybe we should flush first? in case buf got values in it.

Reply:
We always do flush as soon as underlying buffer length equals to `BlockSize`, even if `buf` got values in it.
So in such case `d.len` will hold a value which is `!=0` and another `Write` will add required amount of bytes to `buffer` and `flush` operation will be performed.
I will think about how to do it more clear and accordingly to `go-ethereum` implementation.

Reply:
And I will add more test cases for such scenarios.